File: android/app/src/main/java/com/darkraha/gldemos/gl/GlTexture.kt
package com.darkraha.gldemos.gl


import android.graphics.Bitmap
import android.graphics.BitmapFactory
import android.opengl.GLES30.*
import android.opengl.GLUtils

class GlTexture(val idTexture: Int, val textureUnit: Int, val textureType: Int, val name: String) {
    constructor(idTexture: Int, name: String) : this(idTexture, GL_TEXTURE0, GL_TEXTURE_2D, name)

    fun dispose() {
        glBindTexture(textureType, 0)
        glDeleteTextures(1, intArrayOf(idTexture), 0)
    }

    class TextureBuilder(val idRes: Int, var mName: String) {

        private val idArray = intArrayOf(0)
        var mIdTexture = 0
        var mTextureUnit = GL_TEXTURE0
        var mTextureType = 0
        var mBorder = 0
        var mWidth = 0
        var mHeight = 0
        var mFormat = GL_RGBA
        var mImageData: Bitmap? = null
        var mGenMipmap = true
        var mClampToEdge = false
        fun genMipmap(v: Boolean): TextureBuilder {
            mGenMipmap = v
            return this
        }

        fun clampToEdge(v: Boolean): TextureBuilder {
            mClampToEdge = v
            return this
        }

        fun build2D(): GlTexture {
            mTextureType = GL_TEXTURE_2D

            glGenTextures(1, idArray, 0)
            mIdTexture = idArray[0]

            glBindTexture(mTextureType, mIdTexture)
            loadTexture(idRes, this)

            GLUtils.texImage2D(GL_TEXTURE_2D, 0, mImageData, 0);


            glTexParameteri(mTextureType, GL_TEXTURE_MIN_FILTER, GL_LINEAR)
            glTexParameteri(mTextureType, GL_TEXTURE_MAG_FILTER, GL_LINEAR)

            if (mGenMipmap) {
                glGenerateMipmap(mTextureType)
                glTexParameteri(mTextureType, GL_TEXTURE_MIN_FILTER, GL_LINEAR_MIPMAP_LINEAR)
            }
            if (mClampToEdge) {
                glTexParameteri(mTextureType, GL_TEXTURE_WRAP_S, GL_CLAMP_TO_EDGE)
                glTexParameteri(mTextureType, GL_TEXTURE_WRAP_T, GL_CLAMP_TO_EDGE)
            }
            mImageData?.recycle()
            glBindTexture(GL_TEXTURE_2D, 0)
            return GlTexture(mIdTexture, mTextureUnit, mTextureType, mName)
        }
    }

    companion object {
        fun newTexture2D(idRes: Int, name: String): GlTexture {
            return TextureBuilder(idRes, name).build2D()
        }


        fun loadTexture(idRes: Int, tbuilder: TextureBuilder) {
            val options = BitmapFactory.Options()
            options.inScaled = false // without pre-scaling

            // Read in the resource
            // on the pixel emulator  a 1024x1024 jpg texture worked successful
            tbuilder.mImageData = BitmapFactory.decodeResource(GlCommon.appContext.resources, idRes, options)
        }
    }
}

File: webgl/source/js/gl/gltexture.js
class GlTexture {
    idTexture
    textureUnit
    textureType
    name

    constructor(idTexture, textureUnit, textureType, name) {
        this.idTexture = idTexture;
        this.textureType = textureType;
        this.textureUnit = textureUnit;
        this.name = name;
    }


    dispose(gl) {
        //  gl.deleteTextures(idTexture);
    }

}


class TextureBuilder {

    gl
    mResPath
    mName
    mIdTexture = 0
    mTextureUnit
    mTextureType = 0
    mBorder = 0
    mWidth = 0
    mHeight = 0
    mFormat
    mImageData
    mGenMipmap = true
    mClampToEdge = false

    constructor(gl, resPath, name) {
        this.gl = gl;
        this.mResPath = resPath;
        this.mName = name;
        this.mTextureUnit = gl.TEXTURE0;
        this.mFormat = gl.RGBA;
    }

    genMipmap(v) {
        this.mGenMipmap = v;
        return this;
    }

    clampToEdge(v) {
        this.mClampToEdge = v;
        return this;
    }

    build2D() {
        const gl = this.gl;
        const ret = GlTexture.newSolidColorTexture(gl);
        const image = new Image();

        image.onload = () => {
            ret.textureType = gl.TEXTURE_2D;
            gl.bindTexture(ret.textureType, ret.idTexture);
            gl.texImage2D(ret.textureType, 0, gl.RGBA, gl.RGBA,
                gl.UNSIGNED_BYTE, image);

            gl.texParameteri(ret.textureType, gl.TEXTURE_MIN_FILTER, gl.LINEAR);
            gl.texParameteri(ret.textureType, gl.TEXTURE_MAG_FILTER, gl.LINEAR);


            if (this.mGenMipmap) {
                gl.generateMipmap(ret.textureType);
                gl.texParameteri(ret.textureType,
                    gl.TEXTURE_MIN_FILTER,
                    gl.LINEAR_MIPMAP_LINEAR);
            }

            if (this.mClampToEdge) {
                gl.texParameteri(ret.textureType, gl.TEXTURE_WRAP_S, gl.CLAMP_TO_EDGE);
                gl.texParameteri(ret.textureType, gl.TEXTURE_WRAP_T, gl.CLAMP_TO_EDGE);
            }


            //gl.bindTexture(ret.textureType, 0);
        };
        image.src = this.mResPath;

        return ret;
    }
}



GlTexture.newTexture2D = function(gl, resPath, name) {
    return new TextureBuilder(gl, resPath, name).build2D();
}


/**
 * 
 * @param {*} gl 
 * @param {*} r [0; 255]
 * @param {*} g [0; 255]
 * @param {*} b [0; 255]
 * @param {*} name 
 * @returns 
 */
GlTexture.newSolidColorTexture = function(gl, r, g, b, name) {
    const idTexture = gl.createTexture();
    gl.bindTexture(gl.TEXTURE_2D, idTexture);

    const level = 0;
    const internalFormat = gl.RGBA;
    const width = 1;
    const height = 1;
    const border = 0;
    const srcFormat = gl.RGBA;
    const srcType = gl.UNSIGNED_BYTE;
    const pixel = new Uint8Array([r, g, b, 255]);
    gl.texImage2D(gl.TEXTURE_2D, level, internalFormat,
        width, height, border, srcFormat, srcType,
        pixel);
    //gl.bindTexture(gl.TEXTURE_2D, 0);
    return new GlTexture(idTexture, gl.TEXTURE0, gl.TEXTURE_2D, name);
}